File: app/src/main/java/com/slukhayka/audiobooks/data/identity/RecoveryCodec.kt
package com.slukhayka.audiobooks.data.identity

import java.util.Base64

/**
 * Spec-40 #276 (t2) — the pure recovery-code codec: the stored credential
 * pair (generated email + password) encoded as `SLK1.<base64url(email:pw)>`.
 *
 * The versioned prefix lets the format evolve without mis-decoding older
 * codes; the separator is `:` which cannot occur inside an email address.
 * A garbage, foreign-prefixed, or implausible code decodes to null — a
 * miss, never a crash (the repo's degrade-never rule).
 */
object RecoveryCodec {

    const val PREFIX = "SLK1"

    private const val MAX_EMAIL_LENGTH = 254
    private const val MAX_PASSWORD_LENGTH = 256

    /** Encodes one credential pair into a copyable recovery code. */
    fun encode(email: String, password: String): String {
        val payload = "$email:$password"
        return PREFIX + "." +
            Base64.getUrlEncoder().withoutPadding().encodeToString(payload.toByteArray(Charsets.UTF_8))
    }

    /**
     * Decodes a recovery code back to (email, password), or null when the
     * input is not a code this version can honour.
     */
    fun decode(code: String): kotlin.Pair<String, String>? {
        if (!code.startsWith("$PREFIX.")) return null
        val payload = try {
            String(Base64.getUrlDecoder().decode(code.substring(PREFIX.length + 1)), Charsets.UTF_8)
        } catch (e: IllegalArgumentException) {
            return null
        }
        val separator = payload.indexOf(':')
        if (separator <= 0 || separator == payload.length - 1) return null
        val email = payload.substring(0, separator)
        val password = payload.substring(separator + 1)
        // Plausibility bounds — a decoded blob that cannot be real
        // credentials is rejected outright instead of failing sign-in later.
        if (email.length > MAX_EMAIL_LENGTH || password.length > MAX_PASSWORD_LENGTH) return null
        return email to password
    }
}

File: web/src/identity/recoveryCodec.ts
/**
 * spec-43/T7 — порт RecoveryCodec (spec-40 #276): SLK1.<base64url(email:password)>
 * Версійний префікс дозволяє еволюцію формату; ':' не буває в email.
 * Сміття → null, ніколи не виключення.
 */
export const RECOVERY_PREFIX = 'SLK1'
const MAX_EMAIL_LENGTH = 254
const MAX_PASSWORD_LENGTH = 256

export function encodeRecoveryCode(email: string, password: string): string {
  const payload = `${email}:${password}`
  const bytes = new TextEncoder().encode(payload)
  let binary = ''
  for (const b of bytes) binary += String.fromCharCode(b)
  const base64 = btoa(binary).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/g, '')
  return `${RECOVERY_PREFIX}.${base64}`
}

export function decodeRecoveryCode(code: string): { email: string; password: string } | null {
  if (!code.startsWith(`${RECOVERY_PREFIX}.`)) return null
  const b64 = code.slice(RECOVERY_PREFIX.length + 1)
  if (b64.length === 0) return null
  let padded = b64.replace(/-/g, '+').replace(/_/g, '/')
  while (padded.length % 4 !== 0) padded += '='
  let payload: string
  try {
    const binary = atob(padded)
    const bytes = Uint8Array.from(binary, (c) => c.charCodeAt(0))
    payload = new TextDecoder().decode(bytes)
  } catch {
    return null
  }
  const sep = payload.indexOf(':')
  if (sep <= 0 || sep === payload.length - 1) return null
  const email = payload.slice(0, sep)
  const password = payload.slice(sep + 1)
  if (email.length > MAX_EMAIL_LENGTH || password.length > MAX_PASSWORD_LENGTH) return null
  return { email, password }
}
